src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6474,9 +6483,25 @@ static NullableFlowState getReferenceConversionResultState(TypeWithAnnotations t
             // Converting to a less-derived type (object, interface, type parameter).
             // If the operand is MaybeNull or MaybeDefault, the result should be
             // MaybeNull (if the target type allows) or MaybeDefault otherwise.
-            static NullableFlowState getBoxingConversionResultState(TypeWithState operandType)
+            static NullableFlowState getBoxingConversionResultState(TypeWithAnnotations targetType, TypeWithState operandType)
             {
-                return getConversionResultState(operandType);
+                var state = operandType.State;
+                if (state == NullableFlowState.MaybeNull)
+                {
+                    var type = operandType.Type;
+                    if (type is null || !type.IsTypeParameterDisallowingAnnotationInCSharp8())
+                    {
+                        return NullableFlowState.MaybeDefault;
+                    }
+                    else if (targetType.NullableAnnotation.IsNotAnnotated() &&
+                        type is TypeParameterSymbol typeParameter1 &&
+                        targetType.Type is TypeParameterSymbol typeParameter2 &&
+                        dependsOnTypeParameter(typeParameter1, typeParameter2, NullableAnnotation.NotAnnotated, out var annotation))
```

Comment:
Can `dependsOnTypeParameter` ever return false here (and above in `getReferenceConversionResultState`)?

Reply:
The `dependsOnTypeParameter()` check can fail in `getReferenceConversionResultState()`, with the following for instance (see `NullabilityOfTypeParameters_171()`):
```C#
#nullable enable
class C
{
    static U F<T, U>(T t)
        where T : C?
        where U : T
    {
        return (U)t;
    }
}
```
I suspect the check here in `getBoxingConversionResultState()` cannot fail, because both the target type and the operand type must be unconstrained type parameters, and we only create boxing conversions for those cases when there is an implicit conversion from operand to target type which requires checking the constraint types.

---
In reply to: [466677704](https://github.com/dotnet/roslyn/pull/46405#discussion_r466677704) [](ancestors = 466677704)